### PR TITLE
improve mindfa str expr + reorder dimensions in TCP properties: hosts…

### PR DESCRIPTION
**Pull request by adisos**
… before paths

Signed-off-by: adisos <adisos@il.ibm.com>

### CODE REVIEW

File: tests/expected_cmdline_output_files/livesim_test_all_dot.dot [hunk 1]
```diff
@@ -19,8 +19,8 @@ digraph {
 	"default/deployment-B(Deployment)" -> "kube-system/kube-dns-livesim(Pod)" [label="UDP 53" color="gold2" fontcolor="darkgreen"]
 	"default/foo-app(Pod)" -> "kube-system/kube-dns-livesim(Pod)" [label="UDP 53" color="gold2" fontcolor="darkgreen"]
 	"default/httpbin(Deployment)" -> "kube-system/kube-dns-livesim(Pod)" [label="UDP 53" color="gold2" fontcolor="darkgreen"]
-	"ingress-controller-ns/ingress-controller-livesim(Pod)" -> "default/foo-app(Pod)" [label="TCP {'dst_ports': '5678', 'paths': '/fo{2}(/[\\--9A-Z_a-z]*)?'}" color="gold2" fontcolor="darkgreen"]
-	"istio-ingressgateway-ns/istio-ingressgateway-livesim(Pod)" -> "default/httpbin(Deployment)" [label="TCP {'dst_ports': '80', 'paths': '/(delay|status)[\\--9A-Z_a-z]*', 'hosts': 'httpbin.example.com'}" color="gold2" fontcolor="darkgreen"]
+	"ingress-controller-ns/ingress-controller-livesim(Pod)" -> "default/foo-app(Pod)" [label="TCP {'dst_ports': '5678', 'paths': '/foo*'}" color="gold2" fontcolor="darkgreen"]
```
Comment (zivnevo) [inline tests/expected_cmdline_output_files/livesim_test_all_dot.dot:22]: Shouldn't `paths` be `/foo/*` here?

File: nca/Parsers/IngressPolicyYamlParser.py [hunk 2]
```diff
@@ -155,9 +155,9 @@ def segregate_longest_paths_and_make_dfa(parsed_paths):
                 path_regex = path_string
             else:
                 if path_string:
-                    path_regex = path_string + '|' + path_string + '/' + allowed_chars + '*'
+                    path_regex = path_string + allowed_chars + '*'
```
Comment (zivnevo) [inline nca/Parsers/IngressPolicyYamlParser.py:158]: I'm not sure this is correct. `/foo` should not match `/foobar`. It should only match paths like `/foo/bar`. See [here](https://kubernetes.io/docs/concepts/services-networking/ingress/#path-types)

**Review by zivnevo — CHANGES_REQUESTED** [review]
Some really nice readability improvements. See some comments below.

File: nca/CoreDS/MinDFA.py [hunk 3]
```diff
@@ -109,6 +112,7 @@ def dfa_from_regex(s, alphabet=None):
         # TODO: currently assuming input str as regex only has '*' operator for infinity
         if '*' not in s:
             res.is_all_words = MinDFA.Ternary.FALSE
+        res.regex_expr = s.replace("[.\\w/\\-]*", "*")
```
[the remark this answers is not in the document]
Comment (adisos) [reply nca/CoreDS/MinDFA.py]: Moved this string to be defined at `MinDFA` class, as cannot have circular import.

File: nca/CoreDS/MinDFA.py [hunk 4]
```diff
@@ -176,14 +181,19 @@ def __str__(self):
         str representation of the language accepted by this DFA:
         - option 1: if language has finite number of words -> return string with all accepted words.
         - option 2 (costly): convert fsm to regex with greenery
+        update: changed option 2 to use instead a string of regex expressions with accumulated operations, from
+        which the object was constructed.
```
[the remark this answers is not in the document]
Comment (adisos) [reply nca/CoreDS/MinDFA.py]: done

File: nca/CoreDS/MinDFA.py [hunk 5]
```diff
@@ -219,6 +229,15 @@ def __or__(self, other):
         res = MinDFA.dfa_from_fsm(fsm_res)
         if res.has_finite_len():
             res.is_all_words = MinDFA.Ternary.FALSE
+        # update regex_expr of the result object
+        if self.regex_expr == other.regex_expr:
+            res.regex_expr = self.regex_expr
+        elif other.contained_in(self):
+            res.regex_expr = self.regex_expr
+        elif self.contained_in(other):
+            res.regex_expr = other.regex_expr
```
[the remark this answers is not in the document]
Comment (adisos) [reply nca/CoreDS/MinDFA.py]: It seems they are, removed them.

File: nca/CoreDS/TcpLikeProperties.py [hunk 6]
```diff
@@ -167,8 +171,8 @@ def get_properties_obj(self):
     def __eq__(self, other):
         if isinstance(other, TcpLikeProperties):
             assert self.base_peer_set == other.base_peer_set
-            res = super().__eq__(other) and self.named_ports == other.named_ports and \
-                self.excluded_named_ports == other.excluded_named_ports
+            res = super().__eq__(
+                other) and self.named_ports == other.named_ports and self.excluded_named_ports == other.excluded_named_ports
```
[the remark this answers is not in the document]
Comment (adisos) [reply nca/CoreDS/TcpLikeProperties.py]: fixed.

File: tests/expected_cmdline_output_files/livesim_test_all_dot.dot (hunk 1 above)
Comment (adisos) [reply tests/expected_cmdline_output_files/livesim_test_all_dot.dot]: right, fixed.

File: nca/Parsers/IngressPolicyYamlParser.py (hunk 2 above)
Comment (adisos) [reply nca/Parsers/IngressPolicyYamlParser.py]: Right, thanks. Added the '/' separator back.

File: nca/Parsers/IngressPolicyYamlParser.py [hunk 7]
```diff
@@ -155,7 +155,7 @@ def segregate_longest_paths_and_make_dfa(parsed_paths):
                 path_regex = path_string
             else:
                 if path_string:
-                    path_regex = path_string + '|' + path_string + '/' + allowed_chars + '*'
+                    path_regex = path_string + '/' + allowed_chars + '*'
```
Comment (zivnevo) [inline nca/Parsers/IngressPolicyYamlParser.py:158]: I believe the `/` should not be mandatory. The spec says `/foo` should match `/foo` even if `pathType` is `Prefix`.
Maybe the regex should be something like `path_string + '(/' + allowed_chars + '*)?'`

File: nca/CoreDS/DimensionsManager.py [hunk 8]
```diff
@@ -24,11 +24,12 @@ class __DimensionsManager:
         def __init__(self):
             # TODO: verify alphabet for regex type dimensions, currently using one default alphabet
             #  currently valid chars are: ['.', '/', '-', 0-9, a-z, A-Z ]
-            self.default_dfa_alphabet_chars = ".\\w/\\-"
-            self.default_dfa_alphabet_str = "[" + self.default_dfa_alphabet_chars + "]*"
+            self.default_dfa_alphabet_str = MinDFA.default_alphabet_regex
+            self.default_dfa_alphabet_chars = MinDFA.default_dfa_alphabet_chars
```
Comment (zivnevo) [inline nca/CoreDS/DimensionsManager.py:28]: Are these two data members still needed? I guess `MinDFA` attributes can be used instead...

File: tests/istio_testcases/expected_output/istio_ingress_test_connectivity_map.txt [hunk 9]
```diff
@@ -1,8 +1,8 @@
 final fw rules for query: connectivity, config: istio-ingress:
 src: 0.0.0.0/0 dst_ns: [default,prod,qa] dst_pods: [*] conn: All connections
 src_ns: [default] src_pods: [*] dst_ns: [default] dst_pods: [istio-ingressgateway] conn: All connections
-src_ns: [default] src_pods: [*] dst_ns: [prod] dst_pods: [ratings-v1-857bb87c57] conn: TCP {'dst_ports': '9080', 'paths': '/reviews/[\\--9A-Z_a-z]*', 'hosts': 'eu.bookinfo.com, uk.bookinfo.com'}
-src_ns: [default] src_pods: [*] dst_ns: [qa] dst_pods: [*] conn: TCP {'dst_ports': '7777', 'hosts': 'eu.bookinfo.com, uk.bookinfo.com'}
+src_ns: [default] src_pods: [*] dst_ns: [prod] dst_pods: [ratings-v1-857bb87c57] conn: TCP {'dst_ports': '9080', 'hosts': 'eu.bookinfo.com, uk.bookinfo.com, reviews.prod.svc.cluster.local', 'paths': '/reviews/*'}
+src_ns: [default] src_pods: [*] dst_ns: [qa] dst_pods: [*] conn: TCP {'dst_ports': '7777', 'hosts': 'eu.bookinfo.com, uk.bookinfo.com, reviews.prod.svc.cluster.local'}
```
Comment (zivnevo) [inline tests/istio_testcases/expected_output/istio_ingress_test_connectivity_map.txt:5]: Why do we have a new host here?

**Review by zivnevo — CHANGES_REQUESTED** [review]

File: nca/Parsers/IngressPolicyYamlParser.py (hunk 7 above)
Comment (adisos) [reply nca/Parsers/IngressPolicyYamlParser.py]: Right, changed to this pattern..
Is it readable enough? Could it help to convert to an expression such as `[Prefix]/foo' or `[Exact]/foo` ?

File: nca/CoreDS/DimensionsManager.py (hunk 8 above)
Comment (adisos) [reply nca/CoreDS/DimensionsManager.py]: Removed from `DimensionsManager`

File: tests/istio_testcases/expected_output/istio_ingress_test_connectivity_map.txt (hunk 9 above)
Comment (adisos) [reply tests/istio_testcases/expected_output/istio_ingress_test_connectivity_map.txt]: Due to a fix that I've made in `GenericIngressLikeYamlParser` : the function `parse_regex_host_value` was not handling correctly the host regex when it was just `*` , and after this fix an additional host from the VirtualService was matched by the hosts regex from the Gateway.

**Pull request merged**